File: packages/octospaces-sdk/src/objects/objects.ts
/**
 * Unified Object model — pure logic over the space object index.
 *
 * A space's contents (rooms, categories, docs, projects, tasks, …) are
 * {@link ObjectNode}s in one union-merged index doc at
 * `spaces/{spaceId}/objects/_index`. THIS module is the pure, testable core:
 * the tree builder + merge-artifact guards, breadcrumbs, ordering, and the node
 * reducers a `store.set` applies.
 *
 * Because the index is union-merged (per-node last-write-wins keyed on `updatedAt`),
 * the tree is eventually consistent — two devices can concurrently produce a cycle
 * or an orphan. The builder below is the single place those are repaired so every
 * consumer renders a well-formed tree.
 *
 * **Transitional bridges** (`objectsToRoomCategories`, `roomKindToSubtype`, …) project
 * the object tree into the legacy `Room`-based shape that apps still speak during their
 * migration onto the object model. They are purely mechanical projections over
 * `ObjectNode` and carry no domain-specific names.
 */
import type { ID, ObjectNode, ObjectType, Room, RoomSubtype } from '../core/types.js';
import { randomId, roomSlug } from '../core/ids.js';

/** The bucket new/unfiled objects land in, and the fallback a deleted category's
 *  objects are reassigned to. */
export const DEFAULT_CATEGORY = 'CHANNELS';

/** Deterministic category-node id from its name, so two devices that concurrently
 *  create the SAME category mint the SAME id → the union-merge dedupes them. */
export const categoryId = (name: string): ID => `cat-${roomSlug(name) || randomId()}`;

/** A node plus its resolved children — the shape a tree view renders. */
export interface ObjectTreeNode extends ObjectNode {
  depth: number;
  children: ObjectTreeNode[];
}

/** Map a legacy {@link Room} `kind` to the unified room {@link RoomSubtype}. */
export function roomKindToSubtype(kind: Room['kind']): RoomSubtype {
  switch (kind) {
    case 'dm': return 'dm';
    case 'automated': return 'automation';
    default: return 'channel';
  }
}

/** Inverse of {@link roomKindToSubtype}. A legacy persisted `'stream'` subtype hits
 *  the `default` and reads back as a plain `'channel'` (normalization). */
export function subtypeToRoomKind(subtype: RoomSubtype | undefined): Room['kind'] {
  switch (subtype) {
    case 'dm': return 'dm';
    case 'automation': return 'automated';
    default: return 'channel';
  }
}

function compareSiblings(a: ObjectNode, b: ObjectNode): number {
  if (a.order !== b.order) return a.order - b.order;
  return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
}

/** The order value for a new node appended after `siblings`. */
export function nextOrder(siblings: ObjectNode[]): number {
  let max = 0;
  for (const s of siblings) if (s.order > max) max = s.order;
  return max + 1;
}

/**
 * Build the render tree from a flat node list, repairing merge artifacts:
 *  - **archived** nodes (and their subtrees) are dropped.
 *  - **orphans** — a `parentId` that is missing or archived — reparent to root.
 *  - **cycles** — a node reachable from itself via `parentId` — reparent to root.
 *  - **siblings** sort by {@link compareSiblings} for cross-device determinism.
 */
export function buildTree(nodes: ObjectNode[], includeArchived = false): ObjectTreeNode[] {
  const live = includeArchived ? nodes : nodes.filter((n) => !n.archived);
  const byId = new Map<ID, ObjectNode>(live.map((n) => [n.id, n]));

  const effectiveParent = (n: ObjectNode): ID | null => {
    if (n.parentId == null) return null;
    if (!byId.has(n.parentId)) return null;
    const seen = new Set<ID>([n.id]);
    let cur: ID | null = n.parentId;
    while (cur != null) {
      if (seen.has(cur)) return null;
      seen.add(cur);
      const parent = byId.get(cur);
      if (!parent) return null;
      cur = parent.parentId;
    }
    return n.parentId;
  };

  const childrenOf = new Map<ID | null, ObjectNode[]>();
  for (const n of live) {
    const p = effectiveParent(n);
    const bucket = childrenOf.get(p) ?? [];
    bucket.push(n);
    childrenOf.set(p, bucket);
  }

  function attach(parent: ID | null, depth: number): ObjectTreeNode[] {
    return (childrenOf.get(parent) ?? [])
      .slice()
      .sort(compareSiblings)
      .map((n): ObjectTreeNode => ({ ...n, depth, children: attach(n.id, depth + 1) }));
  }

  return attach(null, 0);
}

/** The root→node trail (inclusive) for breadcrumbs. Returns `[]` if unknown. */
export function breadcrumbs(nodes: ObjectNode[], id: ID): ObjectNode[] {
  const byId = new Map<ID, ObjectNode>(nodes.map((n) => [n.id, n]));
  const trail: ObjectNode[] = [];
  const seen = new Set<ID>();
  let cur: ID | null = id;
  while (cur != null && byId.has(cur) && !seen.has(cur)) {
    seen.add(cur);
    const node: ObjectNode = byId.get(cur)!;
    trail.unshift(node);
    cur = node.parentId;
  }
  return trail;
}

/** The root→parent trail (EXCLUSIVE of the node itself). */
export function ancestors(nodes: ObjectNode[], id: ID): ObjectNode[] {
  return breadcrumbs(nodes, id).slice(0, -1);
}

/** The ids of a node and its whole subtree (for cascade-archive). */
export function subtreeIds(nodes: ObjectNode[], rootId: ID): Set<ID> {
  const childrenOf = new Map<ID | null, ID[]>();
  for (const n of nodes) {
    const bucket = childrenOf.get(n.parentId) ?? [];
    bucket.push(n.id);
    childrenOf.set(n.parentId, bucket);
  }
  const out = new Set<ID>();
  const walk = (id: ID) => {
    if (out.has(id)) return;
    out.add(id);
    for (const child of childrenOf.get(id) ?? []) walk(child);
  };
  walk(rootId);
  return out;
}

// ── Node reducers (pure: ObjectNode[] → ObjectNode[]) ─────────────────────────

export interface NewObjectInput {
  type: ObjectType;
  subtype?: RoomSubtype;
  parentId?: ID | null;
  title: string;
  emoji?: string;
  automation?: import('../core/types.js').AutomationMeta;
  /** Provide to reuse an id (e.g. a room id derived elsewhere); else minted. */
  id?: ID;
}

/** Append a new node under `parentId` at the end of its sibling order. */
export function addObject(nodes: ObjectNode[], input: NewObjectInput, now: number): { nodes: ObjectNode[]; node: ObjectNode } {
  const parentId = input.parentId ?? null;
  const siblings = nodes.filter((n) => n.parentId === parentId);
  const node: ObjectNode = {
    id: input.id ?? `obj-${randomId()}`,
    type: input.type,
    ...(input.subtype ? { subtype: input.subtype } : {}),
    parentId,
    order: nextOrder(siblings),
    title: input.title,
    ...(input.emoji ? { emoji: input.emoji } : {}),
    updatedAt: now,
    ...(input.automation ? { automation: input.automation } : {}),
  };
  return { nodes: [...nodes, node], node };
}

/** Patch a node's mutable metadata (title/emoji/automation), bumping `updatedAt`. */
export function patchObject(nodes: ObjectNode[], id: ID, patch: Partial<Pick<ObjectNode, 'title' | 'emoji' | 'automation'>>, now: number): ObjectNode[] {
  return nodes.map((n) => (n.id === id ? { ...n, ...patch, updatedAt: now } : n));
}

/** Reparent a node (move in the tree). Rejects making a node its own descendant. */
export function reparentObject(nodes: ObjectNode[], id: ID, parentId: ID | null, now: number): ObjectNode[] {
  if (id === parentId) return nodes;
  if (parentId != null && subtreeIds(nodes, id).has(parentId)) return nodes;
  const siblings = nodes.filter((n) => n.parentId === parentId && n.id !== id);
  return nodes.map((n) => (n.id === id ? { ...n, parentId, order: nextOrder(siblings), updatedAt: now } : n));
}

/** Set explicit sibling order (drag-reorder). */
export function reorderObjects(nodes: ObjectNode[], orderById: Record<ID, number>, now: number): ObjectNode[] {
  return nodes.map((n) => (n.id in orderById ? { ...n, order: orderById[n.id]!, updatedAt: now } : n));
}

/** Cascade-archive a node and its whole subtree (soft delete). */
export function archiveObject(nodes: ObjectNode[], id: ID, now: number): ObjectNode[] {
  const ids = subtreeIds(nodes, id);
  return nodes.map((n) => (ids.has(n.id) ? { ...n, archived: true, updatedAt: now } : n));
}

// ── Transitional bridges (Room-based projections) ─────────────────────────────
// Used while apps migrate their content onto the generic object model. Both apps
// still speak the legacy `Room` type; these projections stay until that migration
// is complete.

/** The category→rooms grouping the legacy UI consumes. */
export interface AdaptedCategory {
  name: string;
  rooms: Room[];
}

/**
 * Project the room/category nodes of an index into the legacy `{ name, rooms }[]`
 * shape that app UIs still consume. Category nodes become buckets; room nodes become
 * {@link Room}s grouped under their parent category (or `fallbackCategory` at root).
 * Returns null when the index holds no room/category nodes yet.
 *
 * @deprecated Use the object tree directly once apps complete their migration.
 */
export function objectsToRoomCategories(nodes: ObjectNode[], spaceId: string, fallbackCategory: string): AdaptedCategory[] | null {
  const live = nodes.filter((n) => !n.archived);
  const cats = live.filter((n) => n.type === 'category').slice().sort(compareSiblings);
  const rooms = live.filter((n) => n.type === 'room');
  if (cats.length === 0 && rooms.length === 0) return null;

  const titleById = new Map<ID, string>(cats.map((c) => [c.id, c.title]));
  const buckets = new Map<string, Room[]>();
  for (const c of cats) buckets.set(c.title, []);

  const toRoom = (n: ObjectNode, category: string): Room => ({
    id: n.id,
    spaceId,
    category,
    name: n.title,
    kind: subtypeToRoomKind(n.subtype),
    ...(n.automation ? { automation: n.automation } : {}),
  });

  for (const n of rooms.slice().sort(compareSiblings)) {
    const category = (n.parentId != null && titleById.get(n.parentId)) || fallbackCategory;
    if (!buckets.has(category)) buckets.set(category, []);
    buckets.get(category)!.push(toRoom(n, category));
  }
  return [...buckets.entries()].map(([name, rs]) => ({ name, rooms: rs }));
}

/**
 * Drop `kind: 'automated'` rooms from a category list (they belong to an Agents
 * view, not the main room list). A category that held only agents is removed too.
 *
 * @deprecated Use the object tree directly once apps complete their migration.
 */
export function excludeAutomatedRooms(categories: AdaptedCategory[]): AdaptedCategory[] {
  return categories
    .map((c) => ({ ...c, rooms: c.rooms.filter((r) => r.kind !== 'automated') }))
    .filter((c, i) => c.rooms.length > 0 || categories[i].rooms.length === 0);
}

// ── Seed: build the initial index nodes for a freshly-created space ────────────

/** A minimal object descriptor the {@link seedIndexNodes} builder turns into nodes. */
export interface SeedRoom {
  id: ID;
  name: string;
  kind: Room['kind'];
  category: string;
}

/**
 * Build the initial `ObjectNode[]` for a brand-new space's index: a `category` node
 * per distinct category and a `room` node per seed object parented under it. Pure +
 * deterministic (category ids via {@link categoryId}).
 */
export function seedIndexNodes(rooms: SeedRoom[], now: number): ObjectNode[] {
  const out: ObjectNode[] = [];
  const catId = new Map<string, ID>();
  let catOrder = 0;
  for (const r of rooms) {
    if (catId.has(r.category)) continue;
    const id = categoryId(r.category);
    catId.set(r.category, id);
    out.push({ id, type: 'category', parentId: null, order: catOrder++, title: r.category, updatedAt: now });
  }
  const orderInCat = new Map<ID, number>();
  for (const r of rooms) {
    const parentId = catId.get(r.category)!;
    const order = (orderInCat.get(parentId) ?? 0) + 1;
    orderInCat.set(parentId, order);
    out.push({ id: r.id, type: 'room', subtype: roomKindToSubtype(r.kind), parentId, order, title: r.name, updatedAt: now });
  }
  return out;
}

File: packages/python/octospaces-sdk/octospaces_sdk/objects/objects.py
"""Generic object-tree model — pure logic over a space's object index.

Mirrors ``packages/ts/octospaces-sdk/src/objects/objects.ts`` exactly.

A space's contents are :class:`ObjectNode` dicts in one union-merged index doc
at ``spaces/{spaceId}/objects/_index``.  This module is the pure, testable
core: tree builder + merge-artifact guards, breadcrumbs, ordering, and node
reducers.

Because the index is union-merged (per-node last-write-wins keyed on
``updatedAt``), the tree is eventually consistent — two devices can
concurrently produce a cycle or an orphan.  ``build_tree`` is the single place
those are repaired.
"""

from __future__ import annotations

from typing import Any, Optional

from octospaces_sdk.core.ids import random_id
from octospaces_sdk.core.types import ID, NodeAccess, ObjectNode, ObjectType


# ── Tree node (render tree shape) ─────────────────────────────────────────────


class ObjectTreeNode(dict):  # type: ignore[type-arg]
    """An ``ObjectNode`` dict augmented with ``depth`` and ``children``."""


def _compare_siblings(a: ObjectNode, b: ObjectNode) -> int:
    oa, ob = a.get("order", 0), b.get("order", 0)  # type: ignore[misc]
    if oa != ob:
        return -1 if oa < ob else 1
    aid, bid = a["id"], b["id"]  # type: ignore[index]
    return -1 if aid < bid else (1 if aid > bid else 0)


def next_order(siblings: list[ObjectNode]) -> int:
    """The order value for a new node appended after *siblings*."""
    max_order = 0
    for s in siblings:
        o = s.get("order", 0)  # type: ignore[misc]
        if o > max_order:
            max_order = o
    return max_order + 1


def build_tree(
    nodes: list[ObjectNode], include_archived: bool = False
) -> list[ObjectTreeNode]:
    """Build the render tree from a flat node list, repairing merge artifacts.

    - **Archived** nodes (and their subtrees) are dropped unless
      *include_archived* is ``True``.
    - **Orphans** — a ``parentId`` that is missing or archived — reparent to
      root (``None``).
    - **Cycles** — a node reachable from itself via ``parentId`` — reparent to
      root.
    - **Siblings** sort by ``(order, id)`` for cross-device determinism.
    """
    live = nodes if include_archived else [n for n in nodes if not n.get("archived")]
    by_id: dict[ID, ObjectNode] = {n["id"]: n for n in live}  # type: ignore[index]

    def effective_parent(n: ObjectNode) -> ID | None:
        pid = n.get("parentId")  # type: ignore[misc]
        if pid is None:
            return None
        if pid not in by_id:
            return None
        seen: set[ID] = {n["id"]}  # type: ignore[index]
        cur: ID | None = pid
        while cur is not None:
            if cur in seen:
                return None
            seen.add(cur)
            parent = by_id.get(cur)
            if parent is None:
                return None
            cur = parent.get("parentId")  # type: ignore[misc]
        return pid

    children_of: dict[ID | None, list[ObjectNode]] = {}
    for n in live:
        p = effective_parent(n)
        bucket = children_of.setdefault(p, [])
        bucket.append(n)

    def attach(parent: ID | None, depth: int) -> list[ObjectTreeNode]:
        bucket = sorted(children_of.get(parent, []), key=lambda n: (n.get("order", 0), n["id"]))  # type: ignore[index]
        result: list[ObjectTreeNode] = []
        for n in bucket:
            node: ObjectTreeNode = dict(n)  # type: ignore[assignment]
            node["depth"] = depth
            node["children"] = attach(n["id"], depth + 1)  # type: ignore[index]
            result.append(node)
        return result

    return attach(None, 0)


def breadcrumbs(nodes: list[ObjectNode], id: ID) -> list[ObjectNode]:
    """Root → node trail (inclusive). Returns ``[]`` if unknown."""
    by_id: dict[ID, ObjectNode] = {n["id"]: n for n in nodes}  # type: ignore[index]
    trail: list[ObjectNode] = []
    seen: set[ID] = set()
    cur: ID | None = id
    while cur is not None and cur in by_id and cur not in seen:
        seen.add(cur)
        node = by_id[cur]
        trail.insert(0, node)
        cur = node.get("parentId")  # type: ignore[misc]
    return trail


def ancestors(nodes: list[ObjectNode], id: ID) -> list[ObjectNode]:
    """Root → parent trail (EXCLUSIVE of the node itself)."""
    return breadcrumbs(nodes, id)[:-1]


def subtree_ids(nodes: list[ObjectNode], root_id: ID) -> set[ID]:
    """The ids of a node and its whole subtree (for cascade-archive)."""
    children_of: dict[ID | None, list[ID]] = {}
    for n in nodes:
        bucket = children_of.setdefault(n.get("parentId"), [])  # type: ignore[misc]
        bucket.append(n["id"])  # type: ignore[index]

    out: set[ID] = set()

    def walk(id: ID) -> None:
        if id in out:
            return
        out.add(id)
        for child in children_of.get(id, []):
            walk(child)

    walk(root_id)
    return out


# ── Node reducers (pure: list[ObjectNode] → list[ObjectNode]) ─────────────────


class NewObjectInput(dict):  # type: ignore[type-arg]
    """Input for :func:`add_object`.  All keys are optional except ``type`` and
    ``title``."""


def add_object(
    nodes: list[ObjectNode],
    input: dict[str, Any],
    now: int,
) -> dict[str, Any]:
    """Append a new node under ``parentId`` at the end of its sibling order.

    Returns ``{"nodes": <new list>, "node": <the created node>}``.
    """
    parent_id: ID | None = input.get("parentId")
    siblings = [n for n in nodes if n.get("parentId") == parent_id]  # type: ignore[misc]
    node: ObjectNode = {
        "id": input.get("id") or f"obj-{random_id()}",
        "type": input["type"],
        "parentId": parent_id,
        "order": next_order(siblings),
        "title": input["title"],
        "updatedAt": now,
    }
    if input.get("emoji"):
        node["emoji"] = input["emoji"]  # type: ignore[typeddict-unknown-key]
    if input.get("meta"):
        node["meta"] = input["meta"]  # type: ignore[typeddict-unknown-key]
    access = input.get("access")
    if access and access != "space":
        node["access"] = access  # type: ignore[typeddict-unknown-key]
    if input.get("enc"):
        node["enc"] = True  # type: ignore[typeddict-unknown-key]
    return {"nodes": [*nodes, node], "node": node}


def patch_object(
    nodes: list[ObjectNode],
    id: ID,
    patch: dict[str, Any],
    now: int,
) -> list[ObjectNode]:
    """Patch mutable metadata (title/emoji/meta/access/enc), bumping ``updatedAt``."""
    return [
        dict(n, **patch, updatedAt=now) if n["id"] == id else n  # type: ignore[index]
        for n in nodes
    ]


def reparent_object(
    nodes: list[ObjectNode],
    id: ID,
    parent_id: ID | None,
    now: int,
) -> list[ObjectNode]:
    """Reparent a node (move in the tree). Rejects making a node its own descendant."""
    if id == parent_id:
        return nodes
    if parent_id is not None and parent_id in subtree_ids(nodes, id):
        return nodes
    siblings = [n for n in nodes if n.get("parentId") == parent_id and n["id"] != id]  # type: ignore[index, misc]
    return [
        dict(n, parentId=parent_id, order=next_order(siblings), updatedAt=now)  # type: ignore[index]
        if n["id"] == id  # type: ignore[index]
        else n
        for n in nodes
    ]


def reorder_objects(
    nodes: list[ObjectNode],
    order_by_id: dict[ID, int],
    now: int,
) -> list[ObjectNode]:
    """Set explicit sibling order (drag-reorder)."""
    return [
        dict(n, order=order_by_id[n["id"]], updatedAt=now)  # type: ignore[index]
        if n["id"] in order_by_id  # type: ignore[index]
        else n
        for n in nodes
    ]


def archive_object(nodes: list[ObjectNode], id: ID, now: int) -> list[ObjectNode]:
    """Cascade-archive a node and its whole subtree (soft delete)."""
    ids = subtree_ids(nodes, id)
    return [
        dict(n, archived=True, updatedAt=now) if n["id"] in ids else n  # type: ignore[index]
        for n in nodes
    ]
